esi/lib/node.py

```diff
@@ -0,0 +1,141 @@
+
+import concurrent.futures
+
+from esi import utils
+
+
+def network_list(connection, filter_node=None, filter_network=None):
+
+    def get_baremetal_info(filter_node):
+        baremetal_ports = None
+        baremetal_nodes = None
+
+        if filter_node:
+            baremetal_nodes = [connection.baremetal.find_node(name_or_id=filter_node,
+                                                              ignore_missing=False)]
+            baremetal_ports = [connection.list_nics_for_machine(uuid=baremetal_nodes[0].id)]
+        else:
+            with concurrent.futures.ThreadPoolExecutor() as executor:
+                f1 = executor.submit(connection.list_nics)
+                f2 = executor.submit(connection.list_machines)
+                baremetal_ports = f1.result()
+                baremetal_nodes = f2.result()
+        return baremetal_ports, baremetal_nodes
+
+    def get_network_info(filter_network):
+        if filter_network:
+            filter_network = connection.network.find_network(
+                name_or_id=filter_network,
+                ignore_missing=False
+            )
+
+        with concurrent.futures.ThreadPoolExecutor() as executor:
+            f1 = executor.submit(utils.get_ports, connection, filter_network)
+            f2 = executor.submit(connection.list_networks)
+            f3 = executor.submit(connection.network.ips)
+            network_ports = f1.result()
+            networks_dict = {network.id: network for network in f2.result()}
+            floating_ips = list(f3.result())
+
+        # update floating IP list to include port forwarding information
+        for fip in floating_ips:
+            # no need to do this for floating IPs associated with a port,
+            # as port forwarding is irrelevant in such a case
+            if not fip.port_id:
+                pfws = list(connection.network.port_forwardings(fip))
+                if len(pfws):
+                    fip.port_id = pfws[0].internal_port_id
+                    pfw_ports = ["%s:%s" % (pfw.internal_port,
+                                            pfw.external_port)
+                                 for pfw in pfws]
+                    fip.floating_ip_address = "%s (%s)" % (
+                        fip.floating_ip_address, ','.join(pfw_ports))
+
+        return network_ports, networks_dict, floating_ips
+
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        f1 = executor.submit(get_baremetal_info, filter_node)
+        f2 = executor.submit(get_network_info, filter_network)
+        baremetal_ports, baremetal_nodes = f1.result()
+        network_ports, networks_dict, floating_ips = f2.result()
+
+    data = []
+    for baremetal_node in baremetal_nodes:
+
+        node_ports = []
+        for baremetal_port in baremetal_ports:
+            if baremetal_port.node_id != baremetal_node.id:
+                continue
+
+            network_port_id = baremetal_port.internal_info['tenant_vif_port_id']
```

Comment:
I got an error here when I tested the code. It's because a baremetal port doesn't *have* to have an associated neutron port. If it does, it means the user has associated it with a network; but in its base state it will be unassociated.

Reply:
I think I can just change this to `internal_info.get('tenant_vif_port_id', None)`

Reply:
yep! that's a good suggestion